include/nlohmann/detail/output/binary_writer.hpp

```diff
@@ -676,26 +676,231 @@ class binary_writer
+            oa->write_characters(
+                reinterpret_cast<const CharType*>(name.c_str()),
+                name.size() + 1u);
+
+            write_number<std::int64_t, true>(static_cast<std::int64_t>(j.m_value.number_integer));
+
+            return /*id*/ 1ul + name.size() + 1ul + sizeof(std::int64_t);
+        }
+    }
+
+    std::size_t write_bson_object_internal(const typename BasicJsonType::string_t& name, const BasicJsonType& j)
+    {
+        oa->write_character(static_cast<CharType>(0x03)); // object
+        oa->write_characters(
+            reinterpret_cast<const CharType*>(name.c_str()),
+            name.size() + 1u);
+
+        auto const embedded_document_size = write_bson_object(j);
+
+        return /*id*/ 1ul + name.size() + 1ul + embedded_document_size;
+    }
+
+    std::size_t write_bson_array(const typename BasicJsonType::string_t& name, const BasicJsonType& j)
+    {
+        oa->write_character(static_cast<CharType>(0x04)); // object
+        oa->write_characters(
+            reinterpret_cast<const CharType*>(name.c_str()),
+            name.size() + 1u);
+
+
+        auto document_size_offset = oa->reserve_characters(4ul);
+        std::int32_t embedded_document_size = 5ul;
+
+        for (const auto& el : *j.m_value.array)
+        {
+            embedded_document_size += write_bson_object_entry("", el);
+        }
+
+        oa->write_character(static_cast<CharType>(0x00));
+        write_number_at<std::int32_t, true>(document_size_offset, embedded_document_size);
+
+        return /*id*/ 1ul + name.size() + 1ul + embedded_document_size;
+    }
+
+    std::size_t write_bson_object_entry(const typename BasicJsonType::string_t& name, const BasicJsonType& j)
+    {
+        switch (j.type())
+        {
+            // LCOV_EXCL_START
+            default:
+                assert(false);
+                break;
+            // LCOV_EXCL_STOP
+            case value_t::object:
+                return write_bson_object_internal(name, j);
+            case value_t::array:
+                return write_bson_array(name, j);
+            case value_t::boolean:
+                return write_bson_boolean(name, j);
+            case value_t::number_float:
+                return write_bson_double(name, j);
+            case value_t::number_integer:
+                return write_bson_integer(name, j);
+            case value_t::number_unsigned:
+                return write_bson_unsigned(name, j);
+            case value_t::string:
+                return write_bson_string(name, j);
+            case value_t::null:
+                return write_bson_null(name, j);
+        };
+
+        return 0ul;
+    }
+
+    /*!
+    @param[in] j  JSON value to serialize
+    @pre       j.type() == value_t::object
+    */
+    std::size_t write_bson_object(const BasicJsonType& j)
+    {
+        assert(j.type() == value_t::object);
+        auto document_size_offset = oa->reserve_characters(4ul);
```

Comment:
Could you please document the "magic numbers" like `4ul` or `5ul` below? I think a line in the comment for the function would be sufficient to understand.